Change Tenant<>VMT relationship from 1:n to m:n

nowakowski opened the pull request:
It's a beautiful disaster.

Views will need to be tested manually. Note that VM instantiation logic will remain deficient until such time as it becomes possible to identify the current User/Tenant context (i.e. we need something like @current_user.get_current_tenant, or perhaps simply @current_tenant, or maybe @appliance.appliance_set.user.current_tenant in VM-related services). Please make sure to review the appropriate TODOs that have been left in the code.

nowakowski commented:
I hate that dog. :anger:

nowakowski commented on spec/support/matchers.rb:
```diff
@@ -323,9 +323,9 @@
       actual['appliance_type_id'] == expected.appliance_type_id &&
-      actual['architecture'] == expected.architecture.to_s
+      actual['architecture'] == expected.architecture.to_s &&
+        (expected.tenants.collect{|t| t.id}.include? actual['compute_site_id'])
```
Wow, Hound jackpot - six issues in one line. ;)

mkasztelnik replied:
Here you can use `pluck` as well.

mkasztelnik commented on app/models/atmosphere/virtual_machine_template.rb:
```diff
@@ -71,39 +70,48 @@ class VirtualMachineTemplate < ActiveRecord::Base
+    scope :on_tenant_with_src, ->(t_id, source_id) do
+      joins(:tenants)
+        .where("atmosphere_tenants.tenant_id = '#{t_id}' AND
+        id_at_site = '#{source_id}'")
```
SQL injection is here

nowakowski replied:
Fixed.

nowakowski replied:
Changed to `tenant_id = ?', t_id (...)` syntax which prevents SQL injection.

mkasztelnik commented on app/models/atmosphere/virtual_machine_template.rb:
```diff
@@ -71,39 +70,48 @@ class VirtualMachineTemplate < ActiveRecord::Base
     def uuid
-      "#{tenant.tenant_id}-tmpl-#{id_at_site}"
+      if tenants.blank?
+        "NOTENANT-tmpl-#{id_at_site}"
```
We should avoid something like this. Instead I propose to add validation which checks that at least one tenant is assigned to VMT.

nowakowski replied:
I will keep this assignment as a "goalkeeper" solution in case the database is in an inconsistent state but I will also add the requested validation.

mkasztelnik commented on app/models/atmosphere/virtual_machine_template.rb:
```diff
@@ -71,39 +70,48 @@ class VirtualMachineTemplate < ActiveRecord::Base
       vm_template.name = tmpl_name
       vm_template.managed_by_atmosphere = true
       vm_template.state = :saving
+      vm_template.tenants = [t]
```
Maybe a little bit better?

```ruby
vm_template.tenants << t
```

nowakowski replied:
No, this would automatically trigger a `save` operation on `vm_template` and I don't want to do that just yet.

mkasztelnik commented on app/providers/atmosphere/optimization_strategy/default.rb:
```diff
@@ -33,12 +33,13 @@ def can_scale_manually?
-          tenant_id:
-            appliance.tenants.active.funded_by(appliance.fund)
         )
+        vmts.select do |vmt|
```
If validation checking if VMT is assigned to at least one Tenant will be present this code most probably will not be necessary

nowakowski replied:
Can't hurt to re-check. This is not a performance sink.

mkasztelnik commented on db/migrate/20150701173332_extend_vmt_to_tenant_relationship.rb:
```diff
@@ -0,0 +1,24 @@
+
+  def down
+    # Despawn linking table
+    drop_table :atmosphere_virtual_machine_template_tenants
```
Existing values from `atmosphere_virtual_machine_template_tenants` are not migrated into `tenant_id`, field. First value should be migrated there.

nowakowski replied:
Note that my migration keeps `atmosphere_virtual_machine_templates.tenant_id` so this reassignment is not necessary. This is done in order to ensure a safe transition to multitenancy without irretrievably destroying important data. The column can be removed at a later date.

mkasztelnik commented on spec/requests/atmosphere/api/virtual_machine_flavors_spec.rb:
```diff
@@ -143,13 +143,13 @@
           fl = flavors.first
           expect(fl['memory']).to be >= required_mem
-          expect(fl['compute_site_id']).to eq tmpl.tenant_id
+          expect(tmpl.tenants.collect{|t| t.id}).to include fl['compute_site_id']
```
Using `pluck` you can execute only one SQL query without need of traversing through collection:

```ruby
tmpl.tenants.pluck(:id)
```

nowakowski replied:
Roger.

mkasztelnik commented on spec/support/matchers.rb:
```diff
@@ -260,8 +260,8 @@
-      actual.tenant == template.tenant &&
-      actual.source_template == template
+      actual.source_template == template &&
+        (template.tenants.include? actual.tenant)
```
```ruby
template.tenants.include?(actual.tenant)
```
Maybe a little bit more user friendly?

mkasztelnik commented:
I was trying to start it, unfortunately following exception is thrown in the worker:

```
08:47:01 worker.1 | 2015-07-06T06:47:01.437Z 11527 TID-orovldatg WARN: PG::UndefinedColumn: ERROR:  column atmosphere_virtual_machine_templates.tenant does not exist
08:47:01 worker.1 | LINE 1: ..._virtual_machine_templates"."id_at_site" = $1 AND "atmospher...
08:47:01 worker.1 | : SELECT  "atmosphere_virtual_machine_templates".* FROM "atmosphere_virtual_machine_templates" WHERE "atmosphere_virtual_machine_templates"."id_at_site" = $1 AND "atmosphere_virtual_machine_templates"."tenant" = 1 LIMIT 1
...
08:47:01 worker.1 | /home/marek/git/atmosphere/app/services/atmosphere/vm_updater.rb:120:in `source_template'
```

mkasztelnik commented on app/models/atmosphere/appliance.rb:
```diff
@@ -157,8 +157,10 @@ def billing_state
+      # This method should be overridden in any subproject which makes use of tenants.
       funds = appliance_type.virtual_machine_templates.map do |vmt|
-        vmt.tenant.funds
+        vmt.tenants.first.funds
```
```ruby
vmt.tenants.map(&:funds)
```

Not sure if `&:funds` will work here. 

Add `compact` for `nil` values.

nowakowski replied:
Fixed.

nowakowski commented:
Merge on hold pending resolution of VMTMonitoringWorker issues.

nowakowski commented:
OK, I think I may have caught the problem - VMTUpdater was erroneously duplicating VMT records. Fixed and added spec to prevent this from happening in the future. @mkasztelnik - before you run any more tests you'll need to wipe your local DB (it will have been left in an inconsistent state by the faulty updater).

Btw., caught you pushing a failing build! ;)

mkasztelnik commented:
Please let me know when it will be ready for retests

mkasztelnik commented on app/workers/atmosphere/vm_template_monitoring_worker.rb:
```diff
@@ -36,13 +36,20 @@ def update_images(tenant, images)
       all_tenant_templates.each do |vmt|
-        vmt.destroy(false) if vmt.old?
+        # Unlink VMT from current tenant first
+        tenant.virtual_machine_templates = tenant.virtual_machine_templates-[vmt]
```
This should be done only for "old" VMT

nowakowski replied:
Yep, good catch :clap: 
I have also determined that it's not necessary to invoke `tenant.save` and `vmt.reload`.

mkasztelnik commented on db/migrate/20150701173332_extend_vmt_to_tenant_relationship.rb:
```diff
@@ -0,0 +1,33 @@
+    # Rewrite existing data
+    existing_templates = execute("SELECT id, tenant_id FROM atmosphere_virtual_machine_templates")
+    existing_templates.each do |t|
+      puts "Processing VMT #{t['id'].to_s} which belongs to tenant #{t['tenant_id'].to_s}"
```
`to_s` is not needed here

nowakowski replied:
Aight.

mkasztelnik commented on db/migrate/20150701173332_extend_vmt_to_tenant_relationship.rb:
```diff
@@ -0,0 +1,33 @@
+    existing_templates.each do |t|
+      puts "Processing VMT #{t['id'].to_s} which belongs to tenant #{t['tenant_id'].to_s}"
+      execute("INSERT INTO atmosphere_virtual_machine_template_tenants(virtual_machine_template_id, \
+        tenant_id) VALUES(#{t['id'].to_s}, #{t['tenant_id'].to_s})")
```
Is it possible to convert it into sql save version (query with `?` and parameters given as parameters)?

nowakowski replied:
What's an "sql save version"? ;)

Seriously, though: `execute` accepts a string argument and is not aware of any rails-specific interpolation mechanisms. I take it you're worried about SQL injection but we are dealing **numerical identifiers** which I have only just read from **our very own database**. What you propose could be done with `ActiveRecord::Sanitization.sanitize_sql` but I respectfully refuse to do it as I believe this is pure tinfoil-hat paranoia and needless obfuscation of code. ;)

mkasztelnik commented on db/migrate/20150701173332_extend_vmt_to_tenant_relationship.rb:
```diff
@@ -0,0 +1,33 @@
+    ts = execute("SELECT virtual_machine_template_id AS vmid, MIN(tenant_id) AS tid FROM \
+      atmosphere_virtual_machine_template_tenants GROUP BY(virtual_machine_template_id)")
+    ts.each do |t|
+      execute("UPDATE atmosphere_virtual_machine_templates SET tenant_id = #{t['tid']} WHERE id = #{t['vmid']}")
```
The same comment as above

nowakowski replied:
Nah. :grin:

mkasztelnik commented on db/migrate/20150701173332_extend_vmt_to_tenant_relationship.rb:
```diff
@@ -0,0 +1,33 @@
+    # Recreate tenant_id in vmts
+    add_column :atmosphere_virtual_machine_templates, :tenant_id, :integer
+    # Populate field
+    ts = execute("SELECT virtual_machine_template_id AS vmid, MIN(tenant_id) AS tid FROM \
```
Just a remark (I'm not forcing to change it into such construction). You can use something like:

```ruby
<<-SQL
        atmosphere_appliances.id NOT IN (
          SELECT DISTINCT(appliance_id)
            FROM atmosphere_dev_mode_property_sets)
SQL
```

nowakowski replied:
I don't usually use here documents but okay...

mkasztelnik commented on spec/factories/virtual_machine_templates.rb:
```diff
@@ -1,10 +1,11 @@
     id_at_site { FFaker::Internet.ip_v4_address }
     state :active
 
+    tenants { [FactoryGirl.create(:tenant)] }
```
`build` instead of `create` should be enough there. The reason for it is that some tests are not actually create object in DB only as AR objects.

nowakowski replied:
Okay.

mkasztelnik commented on spec/models/atmosphere/virtual_machine_template_spec.rb:
```diff
@@ -50,6 +50,16 @@
+      expect(vmt.valid?).to be_truthy
+      vmt.tenants = []
+      expect(vmt.valid?).to be_falsy
+      expect(vmt.errors[:tenants]).to include 'A Virtual Machine Template must be attached to at least one Tenant'
```
Why valid true is checked in this spec? Is it not enough to change it into:

```ruby
vmt = build(:virtual_machine_template, tenants: [])
expect(vmt).to_not be_valid
expect(vmt.errors[:tenants]).to include 'A Virtual Machine Template must be attached to at least one Tenant'
```

?

nowakowski replied:
Fixed.

nowakowski commented:
@mkasztelnik You think I can merge it now or are there outstanding issues?
